feat(efb): added functionality to ground services page

Saschl opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->



## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->

- Hooked up the UI with the corresponding SimVars to make all the basic ground services working.

- Also, the buttons were made larger, visual feedback for click/hover and an indication if a service is active was added as well.

- Pushback has to be called now first and will attach to the plane without moving. Afterwards the direction can be selected.

- Doors can be controlled (currently only fwd-left and aft right doors seem to be animated/controllable)


## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->




Active services:
![image](https://user-images.githubusercontent.com/19493808/110257103-99fc3300-7f9c-11eb-9510-dd67968a10ec.png)


## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->
I created this new PR to include the new SimVar framework that was merged afterwards easier.

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):  saschl#9432

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
Open the ground page and check that every item triggers the correct ground service. 

Notes:
The fuel truck might take a while to appear sometimes but that is not something I can influence I believe.

The activation indication is currently tied to the doors (e.g. Jetway is only shown es active once the door actually opens, not while it is moving). I didn't find another way to solve this yet.

When starting pushback with a direction, it still pushes you straight until you click it a second time. I do set the direction but it's not being set until the aircraft actually moves

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

myselfaaryan commented on src/instruments/src/EFB/Ground/Ground.scss:
```diff
@@ -115,6 +115,16 @@
+
+.pushback > div:active,
+.call:active {
+  transform: translateY(2px);
```
@apply translate-y-0.5

St54Kevin commented:
"I have not touched any design stuff for now, please tell me if needed."

Just my thoughts:

- we need bigger buttons in general for services page 
- button color change when pressing and when service is active/inactive
- pushback speed slider

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -16,48 +16,159 @@
+    const [pushBack, setPushBack] = useSimVar('K:TOGGLE_PUSHBACK', 'bool');
+    const [powerActive, setPowerActive] = useSplitSimVar('A:INTERACTIVE POINT OPEN:8', 'percent', 'K:REQUEST_POWER_SUPPLY', 'bool');
+
+    const getTugHeading = (value: number): number => {
+        return (tugHeading  + value) % 360;
+    }
```
```suggestion
    const getTugHeading = (value: number): number => (tugHeading  + value) % 360;
```

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -16,48 +16,159 @@
+     * This ensures the displayed state is in sync with the active services
+     */
+    const applySelectedWithSync = (className: string, id: string, gameSync) => {
+
```
```suggestion
```

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -16,48 +16,159 @@
+        return className;
+    }
+
+    return(
```
```suggestion
    return (
```

Benjozork commented on src/instruments/src/EFB/index-web.tsx:
```diff
@@ -17,9 +17,10 @@
 
 import './Assets/Efb.scss';
+import { render } from '../Common';
+
```
```suggestion
```

theomessin commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -16,48 +16,159 @@
+    const [rampActive, setRampActive] = useSplitSimVar('A:EXIT OPEN:0', 'Enum', 'K:TOGGLE_RAMPTRUCK', 'bool', 500);
+    const [cargoActive, setCargoActive] = useSplitSimVar('A:EXIT OPEN:5', 'Enum', 'K:REQUEST_LUGGAGE', 'bool', 500);
+    const [cateringActive, setCateringActive] = useSplitSimVar('A:EXIT OPEN:3', 'Enum', 'K:REQUEST_CATERING', 'bool', 500);
+    const [fuelingActive, setFuelingActive] = useSplitSimVar('A:INTERACTIVE POINT OPEN:9', 'percent', 'K:REQUEST_FUEL_KEY', 'bool');
+    const [tugHeading, setTugHeading] = useSplitSimVar('PLANE HEADING DEGREES TRUE', 'degrees', 'K:KEY_TUG_HEADING', 'UINT32');
+    const [pushBack, setPushBack] = useSimVar('K:TOGGLE_PUSHBACK', 'bool');
+    const [powerActive, setPowerActive] = useSplitSimVar('A:INTERACTIVE POINT OPEN:8', 'percent', 'K:REQUEST_POWER_SUPPLY', 'bool');
```
That's exactly what this would do!
https://github.com/flybywiresim/a32nx/blob/master/src/instruments/src/Common/simVars.tsx#L270

Saschl replied:
Thanks, I'd still set maxStaleness parameter then? As it's not really that important when/how often the update occurs here.

theomessin replied:
Yeah. The idea is that maxStaleness is a guarantee that we will give you a new value at least every X milliseconds. We might give you more frequent values in the future if the same simVar is being used with a lower maxStaleness. So if you set it to 1000 we will refresh at least once a second.

Benjozork commented on src/instruments/src/EFB/Ground/Ground.scss:
```diff
@@ -59,49 +59,62 @@
 
 .control-grid > div {
   @apply rounded-lg;
```
We should try to move this all to in-html TW in the future. Not necessary for now

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -16,48 +16,198 @@
+
+    /**
+     * allows a direction to be selected directly
+     * rather than fist selected and after that the direction
```
😏 

```suggestion
     * rather than first selected and after that the direction
```

Saschl replied:
:kekw-2:

Benjozork reviewed (approved):
LGTM

Lazarus320 commented:
Category: QA Tester
Name: Lazarus#4286
Date of testing: 1/3/2021
Version of the sim: 1.13.16.0
PR Tested: #3520 
Tier of testing: 2
Changes to observe:

- EFB Ground service functionality 

Testing technique: Loaded c/d at gate and tested the EFB's Ground page 
Steps you've made to test the Pull Request: 
1. Tested GPU 
2. Tested fuel truck
3. Tested baggage truck
4. Tested catering truck
5. Tested jetway connection 
6. Tested pushback 

Flight notes: N/A

Media: N/A

Issues: N/A


Overall Rating: Good 
Conclusions of the testing: All functions worked with no issues good job.

St54Kevin commented:
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 28/02/2021
Version of the sim: 1.13.16.0
PR Tested: #3520 
Tier of testing: 1
Changes to observe: Ground services
Testing technique: 
Steps you've made to test the Pull Request:

1. Loaded at Gate 1 in EDDW
2. Requested all ground services
3. Loaded at a stand 12 in EDDW 
4. Request all ground services. Not all working
5. Loaded into EDDF Gate 40
6. Tested all services
7. Loaded at EDDF Stand V107 (pushback stand). Again only GPU and fuel truck working
8. Tested pushback

Media:

Issues (If Any): When requesting a jetway at the Gate, the Stairs button will also activate (turn white)
When on a stand. Catering, baggage and stairs never arrived.
Pushback truck will immediately start operation (pushing), see suggestion 6.
Pushing at a stand will move the plane without a pusheback truck. To be fair, pushbackhelper also does not send one

Overall Rating: Improvement suggestions below. 

Conclusions of the testing: Needs improvements. I do have some suggestions:

1. A green button for me indicates that the service is already active. I find it a little confusing. Having it grey and turning green if active would be nicer.
2. For services like the fuel truck, baggage, catering that need time (sometimes a lot) to get to the plane, a "requested" label or maybe a blue background until arrived would be nice to see whats already requested. Otherwise, when clicking it again (maybe I cant remember), you would disable it withouth knowing its already on its way. Especially when on a stand. I had no idea wether my request even got acknowledge and is on its way.
Is it possible to get the distance in meters to the approaching service vehicle? We maybe then could display it next to the button
3. Fuel truck at arrival will pop up MSFS fueling window. Maybe coorperate with #3744 to get a link to that page or something else as the MSFS fuel window will get obsolete.
4. As the stairs are not available at a gate, maybe grey it out completely in that case, same for the Jetway on a stand. As the sim knows what to actually pull up, there must be a way to know.
5. The catering button is on the left of the plane. IRL the truck will always come to the right, also in sim it approaches from the right hand side. Having the button on the right too would make sense.
6. I feel like there is a button missing to actually CALL the pushback truck. It will arrive and connect, but will stay stationary. Like its done in pushback helper. Now he immediately starts because we need to push an arrow to get him connect.

Benjozork commented:
@Saschl make this ready for review when you have addressed everything

Saschl commented:
> 
> Issues (If Any):
> 
> * When requesting a jetway at the Gate, the Stairs button will also activate (turn white)
> * When on a stand. Catering, baggage and stairs never arrived.
> * Pushback truck will immediately start operation (pushing), see suggestion 6.
> * Pushing at a stand will move the plane without a pusheback truck. To be fair, pushbackhelper also does not send one
> 
> Overall Rating: Improvement suggestions below.
> 
> Conclusions of the testing: Needs improvements. I do have some suggestions:
> 
> 1. A green button for me indicates that the service is already active. I find it a little confusing. Having it grey and turning green if active would be nicer.
 
 Improved

> 2. For services like the fuel truck, baggage, catering that need time (sometimes a lot) to get to the plane, a "requested" label or maybe a blue background until arrived would be nice to see whats already requested. Otherwise, when clicking it again (maybe I cant remember), you would disable it withouth knowing its already on its way. Especially when on a stand. I had no idea wether my request even got acknowledge and is on its way.

I added a transition phase which turns the button grey until the trigger for that service is true. This phase might be quite short sometimes, as the only way I can currently check is if the door / hose is deployed.

>    Is it possible to get the distance in meters to the approaching service vehicle? We maybe then could display it next to the button

 Good idea, I'm afraid there's no such information available in the SDK, I looked in various places and it's probably not exposed (yet).

> 3. Fuel truck at arrival will pop up MSFS fueling window. Maybe coorperate with #3744 to get a link to that page or something else as the MSFS fuel window will get obsolete.

Good idea, this is possible, however I don't think I can inhibit the MSFS page unfortunately. But the redirect I can addess in another PR.

> 4. As the stairs are not available at a gate, maybe grey it out completely in that case, same for the Jetway on a stand. As the sim knows what to actually pull up, there must be a way to know.

Sadly there's no way. How the ATC handles the stuff is buried inside the sim, I haven't found any reference in the files we have access to. Maybe it will be added with an Update some time I hope.

> 5. The catering button is on the left of the plane. IRL the truck will always come to the right, also in sim it approaches from the right hand side. Having the button on the right too would make sense.

Changed it, however it's funny to see that the truck is technically able to dock on all doors, depending on the position. But 99% of the time it's right.

> 6. I feel like there is a button missing to actually CALL the pushback truck. It will arrive and connect, but will stay stationary. Like its done in pushback helper. Now he immediately starts because we need to push an arrow to get him connect.

Implemented it with one caveat: as it's not possible to check whether the tug is actually attached to the plane (there's a SimVar called `Pushback attached` but that turns true as soon as you call the tug..) you have to wait until it's completely attached to the plane. Otherwise the plane will start moving by itself when hitting the arrow buttons ;) Same behaviour is in pushbackhelper as well

I'm afraid that's most of the stuff I can do with the possibilities the sim gives us at the moment.

For a future PR I'll plan to implement a slider for speed, the ability to "pull" the plane forward and maybe a link to the fuel page :)

Saschl commented:
@Benjozork I think we're good to go now, will update the description in a minute

myselfaaryan reviewed (approved):
LGTM!

TheSquirtingPantin commented:
Category: QA Trial
Name: Discord User Rémi#8871
Date of testing: 10/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3520
Tier of testing: 2
Changes to observe:
Open the ground page and check that every item triggers the correct ground service.

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Start C/D on remote stand (no jetway), EXT PWR and APU on.
2. Open/close doors. Request/Deny services
3. Check of combinations (i.e: Pax Stairs 1L + Opening/closing 1L) 
4. Overhead panel accuracy (EXT PWR) 

Flight notes: N/A



Media: N/A

Issues (If Any): 
 - Door FWD pushbutton issue: Unable to open/close the door when the PAX STAIRS are attached to the plane.Expected behavior: Separation of stairs and doors operation.
 - Door FWD function is inactive/NA after one cycle of PAX STAIRS connecting/disconnecting. 
 - If DOOR FWD is open and PAX STAIRS connected, closing the door with the "Door Fwd" pushbutton will also trigger the PAX STAIRS removal. One occurence only. Expected behavior: Separation of stairs and doors operation. 
 - EXT PWR: Disconnecting the GPU does not turn the green AVAIL on the EXT PWR pushbutton. EXT PWR Available no matter if GPU connected or not. 

Overall Rating: To be improved

Conclusions of the testing:
- Improvement of DOOR FWD function (split STAIRS and DOOR operation, allow more than one full cycle...)
- EXT PWR connection to modify the overhead panel EXT PWR green AVAIL indication.

Kacper848 commented:
Would it be possible to control the cargo door and the baggage service separately too, like pushback helper does it?

Saschl commented:
Hey :)

> * Door FWD pushbutton issue: Unable to open/close the door when the PAX STAIRS are attached to the plane.Expected behavior: Separation of stairs and doors operation.

 This is intentional, controlling the door while the service is active has some weird side effects as MSFS and the EFB are trying to control the door, especially the indication whether a service is active or not. I see it's not ideal but I would like to leave it like this for now.

> * Door FWD function is inactive/NA after one cycle of PAX STAIRS connecting/disconnecting.

 This is probaby an issue. I will investigate.

> * If DOOR FWD is open and PAX STAIRS connected, closing the door with the "Door Fwd" pushbutton will also trigger the PAX STAIRS removal. One occurence only. Expected behavior: Separation of stairs and doors operation.

  This is probably due to the fact that the stairs remove themselves after some time. I don't think I can change anything about that. The door button has no control over the stairs.

> * EXT PWR: Disconnecting the GPU does not turn the green AVAIL on the EXT PWR pushbutton. EXT PWR Available no matter if GPU connected or not.

  Sim limitation iirc. The indication EXT-Power is not tied to the actual GPU that is available. You will notice this on nearly every stand in the airports. I'll keep this issue in mind in case we get some more access to these information at some point :)

Thanks for the report!!

St54Kevin commented:
A little suggestion for the door buttons. Maybe reduce the width on them so they can be visually seperated from the ground services buttons. Now it looks like there are two ground services per door. I would make them at least 50% less width

Matse08 commented:
Category: QA trial
Matse#7454
Date of testing: 11.03.21
Version of the sim: 1.14.5.0
PR Tested:#3520
Tier of testing:1
Changes to observe: Ground crew
Steps you've made to test the Pull Request:
1. Turned on the EFB
2. went to the ground page
3. Checked front door
4. Checked Jetway
5. Checked Fuel truck 
6. checked ground power unit
7. Checked baggage truck
8. Checked back door
9. Checked catering truck
10. Checked pushback
11. Completed the flight and checked the same at the destination
Flight notes N/A
Media N/A
Issues (If Any):
Overall Rating: Good
Conclusions of the testing:
I think that you need to add one button for the jetway and one for the stairs.
The testing went good but i needed to click the tug button 3 times for the Pushback tug to go away.

ChruutvoLuzi commented:
Category: QA Tester
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 11.3.2021
Version of the sim: 1.14.5.0
PR Tested: #3520 
Tier of testing: 2
Changes to observe: Ground handling page on efb

Testing technique: TryOut
Steps you've made to test the Pull Request:
1. Spawned at LSZH
2. Checked all options and tried them out
3. pushed back
4. taxied to another gate
5. tried docking a gate again (worked perfectly)
6. pushed back and taxied to a ramp
7. tried calling "something" with the pax button

Issues (If Any): 
- Sometimes the Pushback Tug Button would bug out and show the opposite state
- No stair arrived after calling pax on a stand
- after taxiing and closing the door, it was not possible to open the door at another ramp
Steps to achieve the issue:
- taxi to a ramp without an pushback tractor and try to call one.
- try to open the doors after taxiing once

Overall Rating: To be improved
Conclusions of the testing: Overall the PR works really well. There are just a few minor things that need to be added/fixed. I also compiled a wishlist that would be very nice to have if possible
Wishes:
- Taxi forward options
- Stair option
- Front right door
